src/Firely.Fhir.Validation/Impl/ExtensionContextValidator.cs

```
@@ -0,0 +1,176 @@
+using Hl7.Fhir.ElementModel;
+using Hl7.Fhir.Model;
+using Hl7.Fhir.Support;
+using Hl7.FhirPath;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Runtime.Serialization;
+
+namespace Firely.Fhir.Validation;
+
+/// <summary>
+/// An assertion which validates the context in which the extension is used against the expected context.
+/// </summary>
+[DataContract]
+[EditorBrowsable(EditorBrowsableState.Never)]
+#if NET8_0_OR_GREATER
+[System.Diagnostics.CodeAnalysis.Experimental(diagnosticId: "ExperimentalApi")]
+#else
+[System.Obsolete("This function is for evaluation purposes only and is subject to change or removal in future updates. Suppress this diagnostic to proceed.")]
+#endif
+public class ExtensionContextValidator : IValidatable
+{
+    /// <summary>
+    /// Creates a new ExtensionContextValidator with the given allowed contexts and invariants.
+    /// </summary>
+    /// <param name="contexts"></param>
+    /// <param name="invariants"></param>
+    public ExtensionContextValidator(IEnumerable<TypedContext> contexts, IEnumerable<string> invariants)
+    {
+        Contexts = contexts.ToList();
+        Invariants = invariants.ToList();
+    }
+
+    internal List<TypedContext> Contexts { get; }
+
+    internal List<string> Invariants { get; }
+
+    /// <summary>
+    /// Validate input against the expected context and invariants.
+    /// </summary>
+    /// <param name="input"></param>
+    /// <param name="vc"></param>
+    /// <param name="state"></param>
+    /// <returns></returns>
+    public ResultReport Validate(IScopedNode input, ValidationSettings vc, ValidationState state)
+    {
+        if (Contexts.Any(c => c.Type == null))
+        {
+            return new IssueAssertion(Issue.PROFILE_ELEMENTDEF_INCORRECT,
+                    "Extension context type was not set, but a context was defined. Skipping non-invariant context validation")
```

Shouldn't we report this in the constructor? Then we can catch it before we ship the validation to the client.

---

We can still validate invariants. What should we do here?

---

I design principle for the validator is that by the time we are validating, we do not report errors anymore about the quality of the profile itself, otherwise put, we cannot bother the end user (who is merely expecting issues about his/her/their instance) with errors that need to be corrected by the author of the profile.  So, we should report this error during compilation.

What we could do (but I don't know that we did it before), is give a WARNING during compilation (which I don't think we can yet), that without this set, only half of the validation will be done. But for now, we should be strict. An incorrect extension context should be flagged as far as I am concerned.